File: packages/crawler-py/shroodler/extractors/jwt_audit.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import re
from datetime import datetime, timedelta, timezone

from shroodler.models import Finding

_JWT_RE = re.compile(r"\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b")

# Small, fast, common-secret wordlist for HMAC weak-secret cracking. This is
# deliberately short (not a gitleaks/hashcat-scale list) so JWT auditing
# doesn't meaningfully slow down a crawl; it catches the "left the tutorial
# default secret in prod" class of bug, not a real brute force.
_WEAK_SECRETS = [
    "secret",
    "password",
    "123456",
    "changeme",
    "changeit",
    "admin",
    "jwt_secret",
    "jwtsecret",
    "your-256-bit-secret",
    "secretkey",
    "secret_key",
    "key",
    "test",
    "testing",
    "dev",
    "development",
    "production",
    "supersecret",
    "super-secret",
    "mysecret",
    "s3cr3t",
    "qwerty",
    "letmein",
    "shroodler",
    "shhh",
    "topsecret",
    "0000",
    "1234",
    "abc123",
    "",
]

_HMAC_ALGS = {
    "HS256": hashlib.sha256,
    "HS384": hashlib.sha384,
    "HS512": hashlib.sha512,
}

_LONG_EXPIRY = timedelta(days=365)


def _b64url_decode(segment: str) -> bytes:
    padded = segment + "=" * (-len(segment) % 4)
    return base64.urlsafe_b64decode(padded.encode("ascii"))


def find_jwts(text: str) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for m in _JWT_RE.finditer(text or ""):
        tok = m.group(0)
        if tok not in seen:
            seen.add(tok)
            out.append(tok)
    return out


def crack_weak_secret(token: str) -> str | None:
    """Try a small common-secret wordlist against an HS256/384/512 JWT.

    Returns the recovered secret string if one of the candidates reproduces
    the token's signature, else None. Non-HMAC algorithms (RS*/ES*/PS*) and
    malformed tokens are not attempted.
    """
    parts = token.split(".")
    if len(parts) != 3:
        return None
    try:
        header = json.loads(_b64url_decode(parts[0]))
    except Exception:
        return None
    alg = str(header.get("alg", "")).upper()
    digestmod = _HMAC_ALGS.get(alg)
    if digestmod is None:
        return None
    signing_input = f"{parts[0]}.{parts[1]}".encode("ascii")
    try:
        target_sig = _b64url_decode(parts[2])
    except Exception:
        return None
    for candidate in _WEAK_SECRETS:
        mac = hmac.new(candidate.encode("utf-8"), signing_input, digestmod).digest()
        if hmac.compare_digest(mac, target_sig):
            return candidate
    return None


def audit_jwt(token: str, url: str) -> list[Finding]:
    parts = token.split(".")
    if len(parts) != 3:
        return []
    try:
        header = json.loads(_b64url_decode(parts[0]))
    except Exception:
        return []
    try:
        payload = json.loads(_b64url_decode(parts[1]))
    except Exception:
        payload = {}

    findings: list[Finding] = []
    ev = token if len(token) <= 40 else token[:20] + "..." + token[-8:]

    alg = str(header.get("alg", ""))
    if alg.lower() == "none":
        findings.append(
            Finding(
                id="jwt-alg-none",
                severity="critical",
                category="secret",
                url=url,
                description=(
                    "JWT declares \"alg\":\"none\" -- if the server accepts this "
                    "token without verifying a signature, an attacker can forge "
                    "arbitrary claims (full authentication/authorization bypass). "
                    "Try replaying this token with the signature stripped."
                ),
                evidence=ev,
            )
        )

    if isinstance(payload, dict) and "exp" not in payload:
        findings.append(
            Finding(
                id="jwt-missing-exp",
                severity="medium",
                category="secret",
                url=url,
                description="JWT has no \"exp\" claim -- the token never expires.",
                evidence=ev,
            )
        )
    elif isinstance(payload, dict):
        try:
            exp = datetime.fromtimestamp(float(payload["exp"]), tz=timezone.utc)
            if exp - datetime.now(timezone.utc) > _LONG_EXPIRY:
                findings.append(
                    Finding(
                        id="jwt-long-expiry",
                        severity="low",
                        category="secret",
                        url=url,
                        description=(
                            f"JWT expiry is unusually far in the future ({exp.date().isoformat()}); "
                            "a leaked long-lived token stays usable for a long time."
                        ),
                        evidence=ev,
                    )
                )
        except (TypeError, ValueError, OSError, OverflowError):
            pass

    secret = crack_weak_secret(token)
    if secret is not None:
        shown = secret if secret else "(empty string)"
        findings.append(
            Finding(
                id="jwt-weak-secret",
                severity="critical",
                category="secret",
                url=url,
                description=(
                    f"JWT signature was reproduced using a common weak secret ({shown!r}) "
                    "from a short built-in wordlist -- an attacker can forge arbitrary "
                    "tokens for this application."
                ),
                evidence=ev,
            )
        )

    return findings


def audit_text(text: str, url: str) -> list[Finding]:
    findings: list[Finding] = []
    for token in find_jwts(text):
        findings.extend(audit_jwt(token, url))
    return findings

File: packages/crawler-go/internal/extractors/jwt_audit.go
package extractors

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/sha512"
	"crypto/subtle"
	"encoding/base64"
	"encoding/json"
	"hash"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/shroodler/crawler-go/internal/models"
)

var jwtRe = regexp.MustCompile(`\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b`)

// weakJWTSecrets is a small, fast common-secret wordlist for HMAC
// weak-secret cracking -- deliberately short so JWT auditing doesn't
// meaningfully slow a crawl down. Kept in sync with
// packages/crawler-py/shroodler/extractors/jwt_audit.py's _WEAK_SECRETS.
var weakJWTSecrets = []string{
	"secret", "password", "123456", "changeme", "changeit", "admin",
	"jwt_secret", "jwtsecret", "your-256-bit-secret", "secretkey",
	"secret_key", "key", "test", "testing", "dev", "development",
	"production", "supersecret", "super-secret", "mysecret", "s3cr3t",
	"qwerty", "letmein", "shroodler", "shhh", "topsecret", "0000", "1234",
	"abc123", "",
}

const longJWTExpiry = 365 * 24 * time.Hour

func b64urlDecode(s string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(s)
}

// FindJWTs returns every distinct JWT-shaped substring in text, in order
// of first appearance.
func FindJWTs(text string) []string {
	seen := map[string]bool{}
	var out []string
	for _, m := range jwtRe.FindAllString(text, -1) {
		if !seen[m] {
			seen[m] = true
			out = append(out, m)
		}
	}
	return out
}

// CrackWeakJWTSecret tries a small common-secret wordlist against an
// HS256/384/512 JWT's signature. Returns the recovered secret and true if
// found; non-HMAC algorithms and malformed tokens are not attempted.
func CrackWeakJWTSecret(token string) (string, bool) {
	parts := splitJWT(token)
	if parts == nil {
		return "", false
	}
	headerB, err := b64urlDecode(parts[0])
	if err != nil {
		return "", false
	}
	var header struct {
		Alg string `json:"alg"`
	}
	if err := json.Unmarshal(headerB, &header); err != nil {
		return "", false
	}
	var newHash func() hash.Hash
	switch header.Alg {
	case "HS256":
		newHash = sha256.New
	case "HS384":
		newHash = sha512.New384
	case "HS512":
		newHash = sha512.New
	default:
		return "", false
	}
	targetSig, err := b64urlDecode(parts[2])
	if err != nil {
		return "", false
	}
	signingInput := []byte(parts[0] + "." + parts[1])
	for _, candidate := range weakJWTSecrets {
		mac := hmac.New(newHash, []byte(candidate))
		mac.Write(signingInput)
		sum := mac.Sum(nil)
		if subtle.ConstantTimeCompare(sum, targetSig) == 1 {
			return candidate, true
		}
	}
	return "", false
}

func splitJWT(token string) []string {
	var parts []string
	start := 0
	for i, c := range token {
		if c == '.' {
			parts = append(parts, token[start:i])
			start = i + 1
		}
	}
	parts = append(parts, token[start:])
	if len(parts) != 3 {
		return nil
	}
	return parts
}

func jwtEvidence(token string) string {
	if len(token) <= 40 {
		return token
	}
	return token[:20] + "..." + token[len(token)-8:]
}

// AuditJWT decodes a single JWT (without verifying it -- there is nothing
// to verify a forged/unsigned token against) and flags alg:none, missing
// or unusually long expiry, and weak HMAC secrets.
func AuditJWT(token, pageURL string) []models.Finding {
	parts := splitJWT(token)
	if parts == nil {
		return nil
	}
	headerB, err := b64urlDecode(parts[0])
	if err != nil {
		return nil
	}
	var header struct {
		Alg string `json:"alg"`
	}
	_ = json.Unmarshal(headerB, &header)

	payloadB, _ := b64urlDecode(parts[1])
	var payload map[string]any
	_ = json.Unmarshal(payloadB, &payload)

	var out []models.Finding
	ev := jwtEvidence(token)

	if strings.ToLower(header.Alg) == "none" {
		desc := "JWT declares \"alg\":\"none\" -- if the server accepts this " +
			"token without verifying a signature, an attacker can forge " +
			"arbitrary claims (full authentication/authorization bypass). " +
			"Try replaying this token with the signature stripped."
		out = append(out, models.Finding{
			ID: "jwt-alg-none", Severity: "critical", Category: "secret",
			URL: pageURL, Description: desc, Evidence: &ev,
		})
	}

	if payload != nil {
		expRaw, hasExp := payload["exp"]
		if !hasExp {
			desc := "JWT has no \"exp\" claim -- the token never expires."
			out = append(out, models.Finding{
				ID: "jwt-missing-exp", Severity: "medium", Category: "secret",
				URL: pageURL, Description: desc, Evidence: &ev,
			})
		} else if expSeconds, ok := asFloat(expRaw); ok {
			exp := time.Unix(int64(expSeconds), 0).UTC()
			if time.Until(exp) > longJWTExpiry {
				desc := "JWT expiry is unusually far in the future (" +
					exp.Format("2006-01-02") +
					"); a leaked long-lived token stays usable for a long time."
				out = append(out, models.Finding{
					ID: "jwt-long-expiry", Severity: "low", Category: "secret",
					URL: pageURL, Description: desc, Evidence: &ev,
				})
			}
		}
	}

	// Deliberately do NOT include the recovered secret itself: this finding
	// (and its evidence) ends up in JSON/HTML/SARIF/JUnit reports that
	// circulate far more widely than the vulnerable app (tickets, CI logs,
	// archived scans) -- printing a live signing secret there would itself
	// be a leak. Re-run CrackWeakJWTSecret manually if you need the value.
	if secret, ok := CrackWeakJWTSecret(token); ok {
		shown := "(empty string)"
		if secret != "" {
			shown = strconv.Itoa(len(secret)) + "-character common secret"
		}
		desc := "JWT signature was reproduced using a " + shown +
			" from a short built-in wordlist -- an attacker can forge arbitrary " +
			"tokens for this application."
		out = append(out, models.Finding{
			ID: "jwt-weak-secret", Severity: "critical", Category: "secret",
			URL: pageURL, Description: desc, Evidence: &ev,
		})
	}

	return out
}

// AuditJWTsInText finds and audits every JWT-shaped token in text.
func AuditJWTsInText(text, pageURL string) []models.Finding {
	var out []models.Finding
	for _, tok := range FindJWTs(text) {
		out = append(out, AuditJWT(tok, pageURL)...)
	}
	return out
}

func asFloat(v any) (float64, bool) {
	switch t := v.(type) {
	case float64:
		return t, true
	case json.Number:
		f, err := t.Float64()
		return f, err == nil
	default:
		return 0, false
	}
}
